### PR TITLE
feat(addon-editor): support any protocols in `link` component

### PR DESCRIPTION
**Pull request by splincode**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the new behavior?

Closes https://github.com/Tinkoff/taiga-ui/issues/4264

https://user-images.githubusercontent.com/12021443/233443162-8a780ef5-c58a-43bb-b76f-78162ae859c8.mov

### CODE REVIEW

File: projects/addon-editor/components/edit-link/test/edit-link-parse.spec.ts [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        expect(tuiEditLinkParseUrl(`a:b:c:d:e`)).toEqual({
+            prefix: `d:`,
+            path: `e`,
+        });
+    });
+
+    it(`URLs`, () => {
+        expect(tuiEditLinkParseUrl(`https://taiga-ui.dev`)).toEqual({
+            prefix: `https://`,
+            path: `taiga-ui.dev`,
```
Comment (nsbarsukov) [inline projects/addon-editor/components/edit-link/test/edit-link-parse.spec.ts:122]: You have two identical tests in one file.
Is it ok ?

File: projects/cdk/utils/miscellaneous/is-valid-url.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+export function tuiIsValidUrl(url: string): boolean {
+    const pattern = new RegExp(
+        `^([a-zA-Z]+:\\/\\/)?` + // protocol
+            `((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|localhost|` + // domain name
+            `((\\d{1,3}\\.){3}\\d{1,3}))` + // OR IP (v4) address
+            `(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*` + // port and path
+            `(\\?[;&a-z\\d%_.~+=-]*)?` + // query string
+            `(\\#[-a-z\\d_]*)?$`, // fragment locator
```
Comment (nsbarsukov) [inline projects/cdk/utils/miscellaneous/is-valid-url.ts:8]: I can't validate it, I just hope it works :D

**Review by nsbarsukov — APPROVED** [review]

File: projects/cdk/utils/miscellaneous/is-valid-url.ts (hunk 2 above)
Comment (splincode) [reply projects/cdk/utils/miscellaneous/is-valid-url.ts:8]: unit tests for help us)

File: projects/addon-editor/components/edit-link/test/edit-link-parse.spec.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-editor/components/edit-link/test/edit-link-parse.spec.ts:105]: fixed

**Review by vladimirpotekhin — APPROVED** [review]

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import {tuiIsValidUrl} from '@taiga-ui/cdk';
+
+interface TuiEditLinkParsed {
+    prefix: string;
+    path: string;
+}
+
+export function tuiEditLinkParseUrl(url?: string): TuiEditLinkParsed {
+    if (url?.startsWith(TUI_EDITOR_LINK_HASH_PREFIX)) {
+        const [anchor = ``] = url?.split(`#`).slice(-1) ?? [];
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:15]: Why not `TUI_EDITOR_LINK_HASH_PREFIX`? Also if it starts with it, why do we need to split and not just use `.slice(1)`?

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    TUI_EDITOR_LINK_SIMPLE_PROTOCOL_DIVIDER,
+} from '@taiga-ui/addon-editor/constants';
+import {tuiIsValidUrl} from '@taiga-ui/cdk';
+
+interface TuiEditLinkParsed {
+    prefix: string;
+    path: string;
+}
+
+export function tuiEditLinkParseUrl(url?: string): TuiEditLinkParsed {
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:13]: Why not use `url = ''` so we do not need all the `?.` later?

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+        if (!!protocol && !!otherPart) {
+            return {prefix: protocol, path: otherPart};
+        }
+    }
+
+    return {prefix, path: prefix === `` ? url ?? `` : path};
+}
+
+function splitOsiProtocol(url?: string): Array<string | undefined> {
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:35]: url = ''

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            : [``, url];
+    const result = [prefix, path].filter(Boolean);
+
+    return path?.includes(`://`) && result.length > 1 ? splitOsiProtocol(path) : result;
+}
+
+function splitSimpleProtocol(url: string): Array<string | undefined> {
+    const [prefix, path] = url.split(/:/).slice(-2).filter(Boolean);
+
+    if (!!prefix && !!path && !tuiIsValidUrl(url)) {
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:58]: Unnecessary `!!`

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return [`${prefix}:`, path];
+    }
+
+    return [];
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:62]: Could have been a ternary.

File: projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    const [prefix = ``, path = ``] = url?.includes(TUI_EDITOR_LINK_OSI_PROTOCOL_DIVIDER)
+        ? splitOsiProtocol(url ?? ``)
+        : splitSimpleProtocol(url ?? ``);
+
+    if (path.includes(TUI_EDITOR_LINK_SIMPLE_PROTOCOL_DIVIDER)) {
+        const [protocol, otherPart] = splitSimpleProtocol(path);
+
+        if (!!protocol && !!otherPart) {
```
Comment (waterplea) [inline projects/addon-editor/components/edit-link/utils/edit-link-parse-url.ts:27]: I thought we have a linter for unnecessary `!!`

File: projects/cdk/utils/miscellaneous/is-valid-url.ts (hunk 2 above)
Comment (waterplea) [reply projects/cdk/utils/miscellaneous/is-valid-url.ts:8]: Chat GPT to the rescue:
image

**Pull request merged**